File: src/main/java/io/github/tanguygab/advancedbungeeexpansion/spigot/AdvancedBungeeExpansion.java
package io.github.tanguygab.advancedbungeeexpansion.spigot;

import io.github.tanguygab.advancedbungeeexpansion.ServerInfo;
import lombok.Getter;
import lombok.experimental.Accessors;
import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import me.clip.placeholderapi.expansion.Taskable;
import net.md_5.bungee.api.ChatColor;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;

import javax.annotation.Nonnull;
import java.util.*;


public class AdvancedBungeeExpansion extends PlaceholderExpansion implements Taskable {

    protected final String CHANNEL = "advancedbungee:channel";
    protected final Map<String, ServerInfo> servers = new HashMap<>();
    protected ServerInfo currentServer;
    protected boolean loaded;
    @Getter private final String identifier = "advancedbungee";
    @Getter private final String author = "Tanguygab";
    @Getter private final String version = "1.0.2";

    @Getter @Accessors(fluent = true) private final boolean canRegister = Bukkit.getServer().spigot().getConfig().getBoolean("settings.bungeecord",false);

    @Override
    public void start() {
        loaded = false;
        Bukkit.getServer().getMessenger().registerIncomingPluginChannel(getPlaceholderAPI(), CHANNEL,new SpigotListener(this));
    }

    @Override
    public void stop() {
        loaded = false;
        Bukkit.getServer().getMessenger().unregisterIncomingPluginChannel(getPlaceholderAPI(), CHANNEL);
        servers.clear();
    }

    @Override
    public String onRequest(OfflinePlayer player, @Nonnull String params) {
        if (!loaded) return "Loading data...";

        String[] args = params.split("_");
        String[] serverArgs = args[0].split(":");
        String category = serverArgs[0];
        String serverName = serverArgs.length > 1 ? serverArgs[1] : null;
        String result = args.length > 1 ? args[1] : "name";

        return switch (category) {
            case "servers" -> switch (result) {
                case "name" -> String.join(", ", servers.keySet());
                case "count" -> String.valueOf(servers.size());
                case "playercount" -> {
                    int i = 0;
                    for (ServerInfo server : servers.values()) i+=server.getPlayerCount();
                    yield String.valueOf(i);
                }
                case "players" -> {
                    List<String> players = new ArrayList<>();
                    servers.values().forEach(server->players.addAll(server.getPlayers()));
                    yield String.join(", ",players);
                }
                default -> null;
            };
            case "current" -> getServerInfo(currentServer,result);
            case "server" -> getServerInfo(servers.get(serverName),result);
            default -> null;
        };
    }

    private String getServerInfo(ServerInfo server, String result) {
        if (server == null) return "Unknown server";
        return switch (result) {
            case "playercount" -> String.valueOf(server.getPlayerCount());
            case "players" -> String.join(", ",server.getPlayers());
            case "status" -> ChatColor.COLOR_CHAR + (server.isStatus() ? "aOnline" : "cOffline");
            case "motd" -> server.getMotd();
            default -> server.getName();
        };
    }
}

File: src/main/java/io/github/tanguygab/advancedbungeeexpansion/spigot/AdvancedBungeeExpansion.kt
package io.github.tanguygab.advancedbungeeexpansion.spigot

import io.github.tanguygab.advancedbungeeexpansion.ServerInfo
import lombok.Getter
import lombok.experimental.Accessors
import me.clip.placeholderapi.expansion.PlaceholderExpansion
import me.clip.placeholderapi.expansion.Taskable
import net.md_5.bungee.api.ChatColor
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import java.util.function.Consumer
import javax.annotation.Nonnull

class AdvancedBungeeExpansion : PlaceholderExpansion(), Taskable {
    @JvmField
    val CHANNEL: String = "advancedbungeejp:channel"
    @JvmField
    val servers: MutableMap<String?, ServerInfo> = HashMap()
    @JvmField
    var currentServer: ServerInfo? = null
    @JvmField
    var loaded: Boolean = false

    override fun getIdentifier() = "advancedbungee"

    override fun getAuthor() = "Tanguygab"

    override fun getVersion() = "1.0.2"

    @Getter
    @Accessors(fluent = true)
    private val canRegister = Bukkit.getServer().spigot().config.getBoolean("settings.bungeecord", false)

    override fun start() {
        loaded = false
        Bukkit.getServer().messenger.registerIncomingPluginChannel(placeholderAPI, CHANNEL, SpigotListener(this))
    }

    override fun stop() {
        loaded = false
        Bukkit.getServer().messenger.unregisterIncomingPluginChannel(placeholderAPI, CHANNEL)
        servers.clear()
    }

    override fun onRequest(player: OfflinePlayer, @Nonnull params: String): String? {
        if (!loaded) return "ロード中..."

        val args = params.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val serverArgs = args[0].split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val category = serverArgs[0]
        val serverName = if (serverArgs.size > 1) serverArgs[1] else null
        val result = if (args.size > 1) args[1] else "name"

        return when (category) {
            "servers" -> when (result) {
                "name" -> java.lang.String.join(", ", servers.keys)
                "count" -> servers.size.toString()
                "playercount" -> {
                    var i = 0
                    for (server in servers.values) i += server.playerCount
                    i.toString()
                }

                "players" -> {
                    val players: MutableList<String> = ArrayList()
                    servers.values.forEach(Consumer { server: ServerInfo -> players.addAll(server.players ?: listOf()) })
                    java.lang.String.join(", ", players)
                }

                else -> null
            }!!

            "current" -> getServerInfo(currentServer, result)
            "server" -> getServerInfo(servers[serverName], result)
            else -> null
        }
    }

    private fun getServerInfo(server: ServerInfo?, result: String): String {
        if (server == null) return "不明"
        return when (result) {
            "playercount" -> server.playerCount.toString()
            "players" -> java.lang.String.join(", ", server.players)
            "status" -> ChatColor.COLOR_CHAR.toString() + (if (server.status) "aオンライン" else "cオフライン")
            "motd" -> server.motd ?: ""
            else -> server.name
        }
    }
}
